be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala

```diff
@@ -17,18 +17,51 @@ case object CoinValidator extends MessageDataContentValidator {
     rpcMessage.getParamsMessage match {
       case Some(message: Message) =>
         val data: PostTransaction = message.decodedData.get.asInstanceOf[PostTransaction]
-        if (data.transactionId != data.transaction.transactionId) {
-          Right(validationError("incorrect transaction id"))
-        } else if (!data.transaction.checkSignatures()) {
-          Right(validationError("bad signature"))
-        } else {
-          data.transaction.sumOutputs() match {
-            case Left(err) => Right(validationError(err.getMessage()))
-            case Right(_)  => Left(rpcMessage)
-          }
-        }
+
+        runChecks(
+          checkId(rpcMessage, data.transactionId, data.transaction.transactionId, validationError("incorrect transaction id")),
+          checkTransactionSignatures(rpcMessage, data.transaction, validationError("bad signature")),
+          checkSumOutputs(rpcMessage, data.transaction, validationError("bad sumOutput"))
+        )
 
       case _ => Right(validationErrorNoMessage(rpcMessage.id))
     }
   }
+
+  /** @param rpcMessage
+    *   rpc message to validate
+    * @param transaction
+    *   The transaction to be checked
+    * @param error
+    *   the error to forward in case the transaction signatures are not valid.
+    * @return
+    *   : GraphMessage passes the rpcMessages to Left if successful, else right with pipeline error
+    */
+  private def checkTransactionSignatures(rpcMessage: JsonRpcRequest, transaction: Transaction, error: PipelineError): GraphMessage = {
+    if (transaction.checkSignatures())
+      Left(rpcMessage)
+    else {
+      Right(error)
+    }
+  }
+
+  /** @param rpcMessage
+    *   rpc message to validate
+    * @param transaction
+    *   The transaction to be checked
+    * @param error
+    *   the error to forward in case the transaction's sum is not valid.
+    * @return
+    *   GraphMessage passes the rpcMessages to Left if successful, else right with pipeline error
+    */
+  private def checkSumOutputs(rpcMessage: JsonRpcRequest, transaction: Transaction, error: PipelineError): GraphMessage = {
+    def validationError(reason: String): PipelineError = super.validationError(reason, "PostTransaction", rpcMessage.id)
```

Comment:
You do not need to define this function inside. It is already defined in the class (see line 15). We simply redefine the validationError which concerns the message, thus we do it inside the class, because it is used in every validators.

Reply:
Mmh after those changes, I believe that the optimal implementation was to keep the same structure, aka: 
`transaction match {
            case Left(err) => Right(validationError(err.getMessage()))
            case Right(_)  => Left(rpcMessage)`

Sorry, I was not realizing that then you would need to redefine a variable :(
I think keeping it as a match case is also more clear. And then I guess you also do not need to put the `error: PipelineError` in the parameters (Sorry again :( )

Reply:
The code doesn't compile when I delete line 58 (def validationError()...
)

Reply:
Mmmh okeey I see now why it does not compile.. 
Indeed you need to access the `err` to use it inside the `validationError` function. 

For now, the parameter `error: PipelineError` is not used inside your function. Can't you try to give as a parameter the function `validationError`, and then use it inside. So, that you do not need to redefine a local one inside.

Reply:
So, you delete the parameter `error: PipelineError` and give as a parameter `validationError: String => PipelineError` instead.